### DISCUSSION TITLE
Is there a way to listen to device status but not using infinite polling?

### DISCUSSION THREAD

**Discussion by castus** [jasonacox/tinytuya#304] [General]
I have a device whose status I want to know. If the device changes its status, I want to do something. I'm wondering what's the best option with this library? From what I read, there is only constant pooling, which is not the fastest way to know the device status. Is there a more "reactive" way to be notified from the device if the status changes?
Thanks

**Comment by jasonacox** [comment]
Tuya devices notify two ways:

1) Cloud - they push all state changes to the Tuya Cloud so that any cloud automation can pick it up and take action.
2) Local Client - if a client is connected to it and sending keep-alive heartbeats, it will stream state changes to that client. Unfortunately, I don't think that is what you are wanting. There are some projects that listen and push these updates on a MQTT topic (example @mafrosis 's  https://github.com/mafrosis/tinytuya2mqtt).  

Here is a  simple monitor loop if you want to try yourself:

```python
import tinytuya

d = tinytuya.OutletDevice('DEVICEID', address='DEVICEIP', local_key='DEVICEKEY', persist=True, version=3.3)

print(" > Send Request for Status < ")
payload = d.generate_payload(tinytuya.DP_QUERY)
d.send(payload)

print(" > Begin Monitor Loop <")
while(True):
    # See if any data is available
    data = d.receive()
    if data:
         print('Received Payload: %r' % data)

    # Send keyalive heartbeat
    print(" > Send Heartbeat Ping < ")
    payload = d.generate_payload(tinytuya.HEART_BEAT)
    d.send(payload)

```

**Comment by uzlonewolf** [comment]
It depends on the device.  Most devices push async updates on state change, though some only update when polled.

To watch for async updates, start a loop which calls `.receive()`; it will block until it receives something or a timeout occurs.  Note that devices close the connection if they doesn't hear anything from you every 30 seconds, so you do need to send it a heartbeat every now and then (I send them every 8-9 seconds just to make sure a couple dropped packets won't close the connection).  My go-to when testing is similar to the loop @jasonacox posted above:

```python
import time
import tinytuya

d = tinytuya.Device( 'DEVICEID', 'DEVICEIP', local_key='DEVICEKEY', persist=True, version=3.x, connection_timeout=1 )

print( d.status() )

print(" > Begin Monitor Loop <")
pingtime = time.time() + 9
polltime = time.time() + 3

while(True):
    if( pingtime <= time.time() ):
        payload = d.generate_payload(tinytuya.HEART_BEAT)
        data = d.send(payload)
        pingtime = time.time() + 9
        if data:
            print( data )

    # only needed for devices which do not update asynchronously
    #if( polltime <= time.time() ):
    #    polltime = time.time() + 3
    #    print( '================' )
    #    print( d.updatedps([17,18,19,20]) )
    #    print( d.status() )

    data = d.receive()

    if data:
        print( data )
```

**Reply by jasonacox** [reply]
Yep, I'm switching to that. 😉

**Comment by castus** [comment]
Thanks, I think I'll go with MQTT as I've it already set up.

I have another question though. What is the reason for not consistent device status when using tinytuya and Smart Life app? I noticed, that when I'm using `tinytuya scan` I have an _905 error_ for one of my devices but:

- I can see it in iot.tuya.com as _Online_
- _Smart Life_ app displays it as _Online_ and I can control it via the app

**Reply by jasonacox** [reply]
Can you paste the output you see?  Also, try with debug mode:

```bash
python3 -m tinytuya scan -d
```

**Reply by uzlonewolf** [reply]
What type of device is it?

A "no response" error can be any number of things.  The wrong protocol version specified, a wrong local key, a firewall blocking the connection, the device could be in a partially-hung state and need to be rebooted, or it could just be a device which does not respond to status requests (some gateways don't respond unless you specify a sub-device).  Since it is the scanner throwing the error it is unlikely to be the wrong protocol version, so I would reboot the device and try it again.

**Reply by castus** [reply]
This device is a WiFi AC plug. Let me share the image with debug info. You can see that each device has "Failed" status.

![SCR-20230321-hfl](https://user-images.githubusercontent.com/522810/226594568-f68a2b1b-99c3-4969-ac57-d812262c84ef.png)

If I use the previous script, with the loop, I have this error:
```
{'Error': 'Network Error: Device Unreachable', 'Err': '905', 'Payload': None}
```

The confusion I have is that I can actually use Smart Life app to control this device :/

**Reply by jasonacox** [reply]
The Smart Life app connects to the cloud and is not running on the same host as your script.  I know it can be confusing, but the connectivity is different from the app.  However, as @uzlonewolf mentioned, this could be a local key issue.  I would suggest:

* Make sure the computer you are using does not have a firewall or is restricting TCP outbound ports (e.g. 6668)
* Try to run `python3 -m tinytuya wizard` to pull the latest local keys.
* Reset one of the devices to see if it helps.

**Reply by castus** [reply]
Unfortunately, nothing works. The firewall couldn't be an issue, because I also have other devices that work. I suspect this might be connected to a device model. 2 of the 5 devices I have, work, and 3 do not. These three devices are from different vendors. Do you think it can be a reason?

**Reply by jasonacox** [reply]
Thanks @castus - looking again at your debug scan, it shows "Connect Failed" without showing any handshake data so that seems to indicate it can't even open the TCP socket.  Are you able to do this?

```bash
# try to reach via ping
ping 192.168.0.127

# try to open TCP port
telnet 192.168.0.127 6668      
```

**Reply by castus** [reply]
It's not possible.
Ps. I got a message from the manufacturer, that those devices are compatible with Smart Life App, not Tuya. How is it possible that I see them in the Tuya platform then? :)

**Reply by jasonacox** [reply]
Thanks! Can you post a link to the devices for anyone else having the same issue?

**Comment by castus** [comment]
I believe these devices are only sold in Poland:

* Smart Plug LTC LXU407 - won't work
* Smart Socket LTC LXU409 - will work

Thanks for your time Folks, I think I can close that discussion.
